File: web/lib/tradingConstraints.ts
import fs from "node:fs";
import { repoConfigPath } from "./repoConfig";

export interface TradingConstraints {
  schema: string;
  allowedBoards: string[];
  priceLimitFractions: Record<string, number>;
  symbolPrefixes: Record<string, string[]>;
  excludeLimitUp: boolean;
  limitSlack: number;
  maxOneDayReturnToBuy: number;
  initialCapitalYuan: number;
  maxDrawdownPct: number;
  maxSinglePositionPct: number;
  maxSingleThemePct: number;
  maxPositions: number;
  decisionTiming: string;
  executionPrice: "next_open";
  supportedSignalFrequencies: string[];
  intradayExecutionPrice: "next_bar_open";
  tPlusOne: boolean;
  lotSize: number;
  maxOrderBarAmountPct: number;
  minHoldingBars: number;
  rebalanceThresholdPct: number;
}

let cached: TradingConstraints | null = null;

function finiteNumber(value: unknown, label: string, minimum = 0): number {
  const parsed = Number(value);
  if (!Number.isFinite(parsed) || parsed < minimum) {
    throw new Error(`invalid trading constraint ${label}: ${String(value)}`);
  }
  return parsed;
}

function positiveInteger(value: unknown, label: string): number {
  const parsed = finiteNumber(value, label, 1);
  if (!Number.isInteger(parsed)) {
    throw new Error(`invalid trading constraint ${label}: expected integer`);
  }
  return parsed;
}

export function loadTradingConstraints(): TradingConstraints {
  if (cached) return cached;
  const file = repoConfigPath("trading-constraints.json");
  const raw = JSON.parse(fs.readFileSync(file, "utf-8")) as Record<string, any>;
  const boards = raw.boards ?? {};
  const buy = raw.buy_constraints ?? {};
  const risk = raw.risk_management ?? {};
  const execution = raw.execution ?? {};
  const priceLimitFractions = Object.fromEntries(
    Object.entries(boards.price_limit_fractions ?? {}).map(([key, value]) => [
      key,
      finiteNumber(value, `boards.price_limit_fractions.${key}`),
    ]),
  );
  for (const required of ["main", "main_st", "star", "chinext", "bse"]) {
    if (!(required in priceLimitFractions)) {
      throw new Error(`missing trading constraint boards.price_limit_fractions.${required}`);
    }
  }
  const executionPrice = String(execution.execution_price ?? "");
  const intradayExecutionPrice = String(execution.intraday_execution_price ?? "");
  if (executionPrice !== "next_open" || intradayExecutionPrice !== "next_bar_open") {
    throw new Error("unsupported execution price in trading-constraints.json");
  }
  cached = {
    schema: String(raw.$schema ?? ""),
    allowedBoards: [...(boards.allowed ?? [])].map(String),
    priceLimitFractions,
    symbolPrefixes: Object.fromEntries(
      Object.entries(boards.symbol_prefixes ?? {}).map(([key, value]) => [
        key,
        Array.isArray(value) ? value.map(String) : [],
      ]),
    ),
    excludeLimitUp: Boolean(buy.exclude_limit_up),
    limitSlack: finiteNumber(buy.limit_slack, "buy_constraints.limit_slack"),
    maxOneDayReturnToBuy: finiteNumber(
      buy.max_one_day_return_to_buy,
      "buy_constraints.max_one_day_return_to_buy",
    ),
    initialCapitalYuan: finiteNumber(
      risk.initial_capital_yuan,
      "risk_management.initial_capital_yuan",
      1,
    ),
    maxDrawdownPct: finiteNumber(risk.max_drawdown_pct, "risk_management.max_drawdown_pct"),
    maxSinglePositionPct: finiteNumber(
      risk.max_single_position_pct,
      "risk_management.max_single_position_pct",
    ),
    maxSingleThemePct: finiteNumber(
      risk.max_single_theme_pct,
      "risk_management.max_single_theme_pct",
    ),
    maxPositions: positiveInteger(risk.max_positions, "risk_management.max_positions"),
    decisionTiming: String(execution.decision_timing ?? ""),
    executionPrice,
    supportedSignalFrequencies: [...(execution.supported_signal_frequencies ?? [])].map(String),
    intradayExecutionPrice,
    tPlusOne: Boolean(execution.t_plus_1),
    lotSize: positiveInteger(execution.lot_size, "execution.lot_size"),
    maxOrderBarAmountPct: finiteNumber(
      execution.max_order_bar_amount_pct,
      "execution.max_order_bar_amount_pct",
    ),
    minHoldingBars: positiveInteger(
      execution.min_holding_bars,
      "execution.min_holding_bars",
    ),
    rebalanceThresholdPct: finiteNumber(
      execution.rebalance_threshold_pct,
      "execution.rebalance_threshold_pct",
    ),
  };
  if (!cached.schema.endsWith("/v2")) {
    throw new Error(`unsupported trading constraints schema: ${cached.schema}`);
  }
  if (!cached.tPlusOne || !cached.supportedSignalFrequencies.includes("5min")) {
    throw new Error("production constraints must enable T+1 and 5min signal support");
  }
  return cached;
}

export function boardForSymbol(symbol: string): "main" | "star" | "chinext" | "bse" {
  const cfg = loadTradingConstraints();
  const code = symbol.replace(/^(sh|sz|bj)/i, "").replace(/\.(sh|sz|bj)$/i, "");
  for (const board of ["star", "chinext", "bse"] as const) {
    if ((cfg.symbolPrefixes[board] ?? []).some((prefix) => code.startsWith(prefix))) {
      return board;
    }
  }
  return "main";
}

export function configuredPriceLimitFraction(symbol: string, name: string): number {
  const cfg = loadTradingConstraints();
  const board = boardForSymbol(symbol);
  if (board === "main" && /ST/i.test(name)) return cfg.priceLimitFractions.main_st;
  return cfg.priceLimitFractions[board];
}

File: research/ashare_research/trading_constraints.py
"""Unified trading-constraints reader for the Python/Research side.

Reads from config/trading-constraints.json so Web and Research share the
same board rules, buy constraints, and risk parameters.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


def _resolve_config_path() -> Path:
    # research/ashare_research/trading_constraints.py → research → repo root
    return Path(__file__).resolve().parents[2] / "config" / "trading-constraints.json"


@dataclass(frozen=True)
class TradingConstraints:
    allowed_boards: list[str]
    price_limit_fractions: dict[str, float]
    symbol_prefixes: dict[str, list[str]]
    exclude_limit_up: bool
    limit_slack: float
    max_one_day_return_to_buy: float
    initial_capital_yuan: float
    max_drawdown_pct: float
    max_single_position_pct: float
    max_single_theme_pct: float
    max_positions: int
    supported_signal_frequencies: list[str]
    intraday_execution_price: str
    t_plus_1: bool
    lot_size: int
    max_order_bar_amount_pct: float
    min_holding_bars: int
    rebalance_threshold_pct: float
    execution_price: str
    decision_timing: str
    raw: dict[str, Any] = field(default_factory=dict, repr=False)

    @property
    def round_trip_fee_bps(self) -> float:
        """Alias for compatibility, derived from the shared cost model."""
        from .cost_config import load_cost_model

        return load_cost_model().round_trip_bps


_cached: TradingConstraints | None = None


def load_trading_constraints() -> TradingConstraints:
    global _cached
    if _cached is not None:
        return _cached
    data = json.loads(_resolve_config_path().read_text("utf-8"))
    _cached = TradingConstraints(
        allowed_boards=data["boards"]["allowed"],
        price_limit_fractions=data["boards"]["price_limit_fractions"],
        symbol_prefixes=data["boards"]["symbol_prefixes"],
        exclude_limit_up=data["buy_constraints"]["exclude_limit_up"],
        limit_slack=data["buy_constraints"]["limit_slack"],
        max_one_day_return_to_buy=data["buy_constraints"]["max_one_day_return_to_buy"],
        initial_capital_yuan=data["risk_management"]["initial_capital_yuan"],
        max_drawdown_pct=data["risk_management"]["max_drawdown_pct"],
        max_single_position_pct=data["risk_management"]["max_single_position_pct"],
        max_single_theme_pct=data["risk_management"]["max_single_theme_pct"],
        max_positions=data["risk_management"]["max_positions"],
        supported_signal_frequencies=data["execution"][
            "supported_signal_frequencies"
        ],
        intraday_execution_price=data["execution"]["intraday_execution_price"],
        t_plus_1=bool(data["execution"]["t_plus_1"]),
        lot_size=int(data["execution"]["lot_size"]),
        max_order_bar_amount_pct=float(
            data["execution"]["max_order_bar_amount_pct"]
        ),
        min_holding_bars=data["execution"]["min_holding_bars"],
        rebalance_threshold_pct=data["execution"]["rebalance_threshold_pct"],
        execution_price=data["execution"]["execution_price"],
        decision_timing=data["execution"]["decision_timing"],
        raw=data,
    )
    return _cached


def price_limit_fraction(symbol: str, name: str = "") -> float:
    """A-share daily price-limit fraction by board."""
    cfg = load_trading_constraints()
    code = re.sub(r"^(sh|sz|bj)", "", symbol, flags=re.IGNORECASE)
    code = re.sub(r"\.(sh|sz|bj)$", "", code, flags=re.IGNORECASE)
    for board in ("star", "chinext", "bse"):
        prefixes = cfg.symbol_prefixes.get(board, [])
        if any(code.startswith(p) for p in prefixes):
            return cfg.price_limit_fractions[board]
    if re.search(r"ST", name, re.IGNORECASE):
        return cfg.price_limit_fractions["main_st"]
    return cfg.price_limit_fractions["main"]


def is_limit_up(close: float, prev_close: float, symbol: str, name: str = "") -> bool:
    cfg = load_trading_constraints()
    frac = price_limit_fraction(symbol, name)
    if prev_close <= 0:
        return False
    return close / prev_close - 1 >= frac - cfg.limit_slack
